aten/src/ATen/cuda/CUDAStream.cpp

```diff
@@ -1,57 +1,127 @@
 #include "ATen/cuda/CUDAStream.h"
 #include "ATen/cuda/CUDAContext.h"
 #include "ATen/cuda/Exceptions.h"
+#include "ATen/DeviceGuard.h"
 #include "ATen/Error.h"
 
 #include <mutex>
 #include <atomic>
+#include <cstdint>
+#include <deque>
+#include <vector>
 
 // Internal implementation is entirely hidden
+// Note: CUDAStreamInternals doubles for a THCStream
 struct CUDAStreamInternals {
-  bool is_destructible;
-  std::atomic<int> refcount;
-  int64_t device; // Note: cudaGetDevice works with int32_t, not int64_t
-  cudaStream_t stream;
+  CUDAStreamInternals() = default;
+
+  ~CUDAStreamInternals() {
+    if (stream) cudaStreamDestroy(stream);
+  }
+
+  int64_t device = -1; 
+  cudaStream_t stream = nullptr;
 };
 
 namespace at {
 namespace cuda {
 
 namespace detail {
 
-  /*
-  * Stream state
-  */
-  static constexpr cudaStream_t DEFAULT_STREAM = 0;
-
-  static std::once_flag init_flag;
+  // Global stream state and constants
   static int64_t num_gpus;
-  static CUDAStreamInternals* default_streams;
+  static constexpr int STREAMS_PER_POOL = 32;
+  static constexpr unsigned int DEFAULT_FLAGS = cudaStreamNonBlocking;
+  static int HIGH_PRIORITY = 0;
+  static int LOW_PRIORITY = 0;
+  
+  // Default streams
+  static std::once_flag init_flag;
+  static std::vector<CUDAStreamInternals> default_streams;
+
+  // Non-default streams
+  static std::deque<std::once_flag> device_flags;
+  static std::deque<std::atomic<int>> low_priority_counters;
```

Comment:
A short comment saying what the counters count would be appreciated. It's a mapping of device to number of in use streams for that device, if I understand correctly.